STMicroelectronics/STM32L4_L4+/app/az_ulib_dm/src/az_ulib_registry.c

```diff
@@ -0,0 +1,300 @@
+    // HAL_OK)
+    // {
+    //   result = result_from_hal_status(hal_status);
+    //   AZ_ULIB_THROW_IF_AZ_ERROR(result);
+    // }
+    AZ_ULIB_THROW_IF_AZ_ERROR(
+        result_from_hal_status(internal_flash_write_doubleword((uint32_t)address, REGISTRY_READY)));
+  }
+  AZ_ULIB_CATCH(...) {}
+
+  return AZ_ULIB_TRY_RESULT;
+}
+
+static az_result set_registry_node_delete_flag(uint8_t* address)
+{
+  AZ_ULIB_TRY
+  {
+    // if ((hal_status = internal_flash_write_doubleword(
+    //          ((uint32_t)address + AZ_ULIB_REGISTRY_FLAG_SIZE), REGISTRY_DELETED))
+    //     != HAL_OK)
+    // {
+    //   result = result_from_hal_status(hal_status);
+    //   AZ_ULIB_THROW_IF_AZ_ERROR(result);
+    // }
+    AZ_ULIB_THROW_IF_AZ_ERROR(result_from_hal_status(internal_flash_write_doubleword(
+        ((uint32_t)address + AZ_ULIB_REGISTRY_FLAG_SIZE), REGISTRY_DELETED)));
+  }
+  AZ_ULIB_CATCH(...) {}
+
+  return AZ_ULIB_TRY_RESULT;
+}
+
+static az_result store_registry_node(az_ulib_registry_node node, uint8_t** node_ptr)
+{
+  AZ_ULIB_TRY
+  {
+    az_ulib_registry_node* runner;
+    int32_t node_index;
+
+    runner = (az_ulib_registry_node*)(&__REGISTRYINFO);
+
+    /* Look for next available spot */
+    for (node_index = 0; node_index < MAX_AZ_ULIB_REGISTRY_ENTRIES; node_index++)
+    {
+      if (runner->_internal.status.ready_flag == REGISTRY_FREE)
+      {
+        break;
+      }
+      runner++;
+    }
+
+    /* Handle case if all nodes were used */
+    AZ_ULIB_THROW_IF_ERROR((node_index < MAX_AZ_ULIB_REGISTRY_ENTRIES), AZ_ERROR_OUT_OF_MEMORY);
+
+    /* Store pointers to key value pair into flash */
+    AZ_ULIB_THROW_IF_AZ_ERROR(result_from_hal_status(internal_flash_write(
+        (uint8_t*)(runner) + sizeof(az_ulib_registry_node_status), // add size of flags
+        (uint8_t*)&(node._internal.data),
+        sizeof(az_ulib_registry_node_data))));
+
+    /* Return pointer to this node for setting flags later */
+    *node_ptr = (uint8_t*)runner;
+  }
+  AZ_ULIB_CATCH(...) {}
+
+  return AZ_ULIB_TRY_RESULT;
+}
+
+/* Find pointer to the next available space in flash to store registry key/value */
+static uint8_t* flash_buf_get()
+{
+  uint8_t* flash_ptr;
+  az_ulib_registry_node node;
+
+  /* Is there an empty registry? */
+  node = *(az_ulib_registry_node*)(&__REGISTRYINFO);
+  if (node._internal.status.ready_flag == REGISTRY_FREE)
+  {
+    flash_ptr = (uint8_t*)(&__REGISTRY);
+  }
+  else
+  {
+    uint8_t* max_address = (uint8_t*)(&__REGISTRY); /* The beginning of buffer space */
+    az_ulib_registry_node* last_node;
+    az_ulib_registry_node* runner;
+
+    /* Find occupied node that has maximum address such that a new key-value pair can be stored */
+    runner = (az_ulib_registry_node*)(uint8_t*)(&__REGISTRYINFO);
+    for (int i = 0; i < MAX_AZ_ULIB_REGISTRY_ENTRIES; i++)
+    {
+      if (runner->_internal.status.ready_flag == REGISTRY_READY)
+      {
+        if (az_span_ptr(runner->_internal.data.value) > max_address)
```

Comment:
>max_address [](http://example.com/codeflow?start=56&length=11)

You have 2 variables controlling the last address, `last_node` and `max_address`. Consider use `az_span_ptr(last_node->_internal.data.value` instead of create the second control `max_address`. You only need to take care of the first valid node.